### PR TITLE
RATEST-240:Migrate Pages from Refapp distro to Qaframework module

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-240

### CODE REVIEW

**Comment by sherrif10** [comment]
I think  Before deleting the pages from distro, will first work on migrating all the selenium tests into qaframework since pages in distro depend on selenium tests.we wont delete pages and leave selenium tests, we need to work on the both such that distro remains working as expected

**Comment by kdaud** [comment]
@sherrif10 have you ensured that the guys you are migrating in this PR have similar content(updated) with their origin?

File: pom.xml [hunk 1]
```diff
@@ -31,23 +29,50 @@
...
+			<version>2.11.2</version>
+		</dependency>
         <dependency>
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-log4j12</artifactId>
             <version>1.5.6</version>
         </dependency>
+        <dependency>
+		    <groupId>com.saucelabs</groupId>
+		    <artifactId>sauce_junit</artifactId>
```
Comment (kdaud) [inline pom.xml:73]: What is the use of this dependency?

File: src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 2]
```diff
@@ -46,8 +46,7 @@
 	public void visitHomePage() {
 		testPatient = createTestPatient();
 		initiateWithLogin();
-		new TestData.TestVisit(testPatient.uuid, TestData.getAVisitType(),
-				getLocationUuid(homePage)).create();
+		new TestData.TestVisit(testPatient.uuid, TestData.getAVisitType(),getLocationUuid(homePage)).create();
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:49]: Worth adding a space something like;.

```js
new TestData.TestVisit(testPatient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
```

File: src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	// public void search(String text) {
+	// WebElement searchField = findElement(PATIENT_SEARCH);
+	// try {
+	// searchField.sendKeys(text);
+	// } catch(Exception e) {
+	// e.printStackTrace();
+	// }
+	// clickOn(FIRST_FOUND_PATIENT);
+	// }
+}
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java:43]: is CaptureVitalsPage exercised any where?

File: src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+	private static final By PATIENT_GIVENNAME = By.cssSelector("#content div span.PersonName-givenName");
+	private static final By ACTIVE_VISIT_MESSAGE = By.cssSelector("active-visit-message");
+	private static final By ALLERGIES_LINK = By.id("allergyui-editAllergies");
+	private static final By CONDITIONS_LINK = By.cssSelector(".conditions .info-header i.right");
+	private static final By ATTACHMENTS_LINK = By.cssSelector("a[id='attachments.attachments.visitActions.default']");
+	private static final By ACTIVE_VISIT = By.cssSelector("visitbyencountertype ul li a");
+	private static final By DELETE_PATIENT_LINK = By.id("org.openmrs.module.coreapps.deletePatient");
+	private static final By DELETE_PATIENT_REASON = By.id("delete-reason");
+	private static final By DELETE_PATIENT_CONFIRM = By.cssSelector("#delete-patient-creation-dialog > div.dialog-content > button.confirm.right");
+	private static final By DELETE_PATIENT_CANCEL = By.cssSelector("#delete-patient-creation-dialog > div.dialog-content > button.cancel");
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java:87]: I do not see where DELETE_PATIENT_CANCEL is used. Do you really need it?

File: src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+	private static final By EDIT_PATIENT_LINK = By.cssSelector("a[id='application.registrationapp.summary.editPatientLink']");
+	private static final String REASON = "patient nolonger needed";
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java:89]: The same with REASON

File: src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java (hunk 4 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java:87]: These functionalities are used in deletePatientTests

File: src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java (hunk 2 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:49]: Havent seen any difference according to your suggestion.

File: src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java (hunk 4 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java:87]: Is there any function that complains when you comment  DELETE_PATIENT_CANCEL?

File: src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java (hunk 3 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java:43]: Ofcourse its very excersed i do remember you are the one working on capturevitalsPage :smile:

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java:43]: CaptureVitalsPage is not used any where so we need to have it removed among the pages.

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java:43]: I do not see that page used any where. Have you had a look at the pages imported within CaptureVitalsTest?

```js
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.openmrs.reference.groups.BuildTests;
import org.openmrs.reference.page.ActiveVisitsPage;
import org.openmrs.reference.page.ClinicianFacingPatientDashboardPage;
import org.openmrs.reference.page.PatientCaptureVitalsPage;
import org.openmrs.uitestframework.test.TestData;
```

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java:43]: We would need only pages with tests exercising their functionalities.

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/page/CaptureVitalsPage.java]: Resolved thanks

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java [hunk 6]
```diff
@@ -0,0 +1,695 @@
+package org.openmrs.contrib.qaframework.helper;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.ws.rs.NotFoundException;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.*;
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java:9]: Rather import the specific packages required than explicitly calling all the `util.*` packages

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java [hunk 7]
```diff
@@ -0,0 +1,695 @@
...
+		return SINGLETON;
+	}
+
+	public static String getALocation() {
+		JsonNode locations = RestClient.get("location");
+		return locations.get("results").get(0).get("uuid").asText(); // arbitrarily
+																		// choose
+																		// the
+																		// first
+																		// location
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java:48]: Remove all the redundant empty lines in the method `getALocation()`

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+
+	public static String getAVisitType() {
+		JsonNode locations = RestClient.get("visittype");
+		return locations.get("results").get(0).get("uuid").asText(); // arbitrarily
+																		// choose
+																		// the
+																		// first
+																		// visit
+																		// type
+	}
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java:68]: Remove all the redundant empty lines in `getAVisitType()`

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestProperties.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+	public WebDriverType getWebDriver() {
+		try {
+			return WebDriverType.valueOf(getBrowser());
+		} catch (IllegalArgumentException e) {
+			return WebDriverType.firefox;
+		}
+	}
+
+	; // only these two for now
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestProperties.java:120]: Of what value is the semi-colon here?

File: src/test/java/org/openmrs/contrib/qaframework/page/ActiveVisitsPage.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * <p>
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ActiveVisitsPage.java:9]: @sherrif10 I have observed this trade made in many of the OpenMRS repos and I  want to think this is the right way of writing trade mark. 

Would be great to have all the files in this PR having this trade mark and am wondering whether are you able to add this trade mark all through the files apart from pom.xml ?

File: src/test/java/org/openmrs/contrib/qaframework/page/ServicePage.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+package org.openmrs.contrib.qaframework.page;
+
+import org.openmrs.contrib.qaframework.helper.Page;
+import org.openqa.selenium.By;
+
+/**
+ * Created by tomasz on 24.06.15.
+ */
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ServicePage.java:18]: As well the comment here

File: src/test/java/org/openmrs/contrib/qaframework/page/VisitNotePage.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	public void addDiagnosis(String diag) {
+		WebElement diagnosisElement = findElement(DIAGNOSIS_SEARCH);
+		diagnosisElement.click();
+		enterDiagnosis(diag);
+		diagnosisElement.clear();
+		diagnosisElement.click();
+
+	}
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/VisitNotePage.java:58]: Remove the redundant empty line(s) in addDiagnosis();

**Review by kdaud — CHANGES_REQUESTED** [review]
Great work @sherrif10 

Worth looking into the comments have left above!

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestBase.java [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+				"ondemand.saucelabs.com:80");
+
+		if (!StringUtils.isBlank(sauceLabsUsername)
+				&& !StringUtils.isBlank(sauceLabsAccessKey)) {
+			sauceLabsAuthentication = new SauceOnDemandAuthentication(
+					sauceLabsUsername, sauceLabsAccessKey);
+			sauceLabsResultReportingTestWatcher = new SauceOnDemandTestWatcher(
+					this, sauceLabsAuthentication);
+		}
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestBase.java:112]: Remove the redundant empty line here

**Review by kdaud — CHANGES_REQUESTED** [review]
@sherrif10 catch the comments above and also cross check through the files which are missing OpenMRS trade mark and include it before the package import.

Something like ;

```js

/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 * 
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs.contrib.qaframework.*;

**Comment by sherrif10** [comment]
cc @kdaud  any more reviews will be highly welcome

File: src/test/java/org/openmrs/contrib/qaframework/helper/BuildTests.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/BuildTests.java:10]: usually the package declaration comes immediately after the trade mark, something like  
```js
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 * <p>
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs.contrib.qaframework.helper;
```

File: src/test/java/org/openmrs/contrib/qaframework/helper/BuildTests.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+package org.openmrs.contrib.qaframework.helper;
+
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/BuildTests.java:13]: isn't one empty line enough here?

File: src/test/java/org/openmrs/contrib/qaframework/helper/GenericPage.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.openmrs.contrib.qaframework.helper;
+
+import org.openqa.selenium.WebDriver;
+
+/**
+ * A Page that represents any page, i.e. a page that we don't (yet) know which
+ * page it is.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/GenericPage.java:18]: is the empty line necessary here?

File: src/test/java/org/openmrs/contrib/qaframework/helper/Page.java [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+import org.openqa.selenium.support.ui.Select;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import com.google.common.collect.Iterables;
+
+/**
+ * A superclass for "real" pages. Has lots of handy methods for accessing
+ * elements, clicking, filling fields. etc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/Page.java:39]: is the empty line necessary after the comment?

File: src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java [hunk 18]
```diff
@@ -0,0 +1,697 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import jakarta.ws.rs.NotFoundException;
+
+
+public class TestData {
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/helper/TestData.java:28]: @sherrif10 why did you marked the conversation here as resolved But when itsn't?

File: src/test/java/org/openmrs/contrib/qaframework/page/ActiveVisitsPage.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ActiveVisitsPage.java:10]: is the empty line here necessary?

File: src/test/java/org/openmrs/contrib/qaframework/page/AddEditLocationPage.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/AddEditLocationPage.java:10]: is the empty line here necessary?

File: src/test/java/org/openmrs/contrib/qaframework/page/AppointmentSchedulingPage.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/AppointmentSchedulingPage.java:10]: even here

File: src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java:10]: the same here

File: src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java [hunk 23]
```diff
@@ -0,0 +1,462 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ClinicianFacingPatientDashboardPage.java:10]: even here

File: src/test/java/org/openmrs/contrib/qaframework/page/ConditionsPage.java [hunk 24]
```diff
@@ -0,0 +1,109 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+
```
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/page/ConditionsPage.java:10]: even here

**Review by kdaud — APPROVED** [review]
Well done @sherrif10 

Keep an eye on @dkayiwa's final reviews and catch them appropriately to have this daughter safely married to master!

File: src/test/resources/readme.txt [hunk 25]
```diff
@@ -0,0 +1,6 @@
+These chromedriver executables were downloaded from http://chromedriver.storage.googleapis.com/index.html
+
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply src/test/resources/readme.txt:2]: Its migrated from https://github.com/openmrs/openmrs-contrib-uitestframework/blob/master/src/main/resources/readme.txt

**Comment by kdaud** [comment]
Are the guys being deleted in the PR you referenced all being migrated here?

File: src/test/resources/readme.txt (hunk 25 above)
[the remark this answers is not in the document]
Comment (dkayiwa) [reply src/test/resources/readme.txt:2]: How about the rest of the files that you left there and did not copy over to here?

**Comment by kdaud** [comment]
@sherrif10 it logically implies we need to put on hold the other PR not to be merged until all the guys have been migrated safely.

File: src/test/resources/readme.txt (hunk 25 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply src/test/resources/readme.txt:2]: @dkayiwa I seem to have removed every file that we need in uitestframework module, Probably files I left there we have them already in qaframework.

File: src/test/resources/post.js [hunk 26]
```diff
@@ -0,0 +1,24 @@
+// Borrowed from http://stackoverflow.com/a/133997
+function post(path, params, method) {
+    method = method || "post"; // Set method to post by default if not specified.
```
Comment (sherrif10) [inline src/test/resources/post.js:3]: cc @dkayiwa  here is post.js file being migrated

File: src/test/resources/post.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
+// Borrowed from http://stackoverflow.com/a/133997
+function post(path, params, method) {
```
Comment (sherrif10) [inline src/test/resources/post.js:2]: cc @dkayiwa  here is the file .Thanks

**Comment by sherrif10** [comment]
@dkayiwa  Looking forward for more reviews.Thanks

**Comment by sherrif10** [comment]
> Where is the pr for deleting the files you have added here?

https://github.com/openmrs/openmrs-distro-referenceapplication/pull/598

**Pull request merged**